Add AreaBeam target tracking

reaperrr opened the pull request:
Adds support for target tracking to `AreaBeam` and enables it on TS Sonic tank weapon.

Closes #10181.

reaperrr committed:
Add target tracking support to AreaBeam

Closes #10181.

reaperrr committed:
Enable target tracking on TS SonicZap

pchote commented on OpenRA.Mods.Common/Projectiles/AreaBeam.cs:
```diff
@@ -132,8 +135,24 @@ public AreaBeam(AreaBeamInfo info, ProjectileArgs args, Color color)
+		void UpdateTarget()
+		{
+			if (args.GuidedTarget.IsValidFor(args.SourceActor))
+				target = args.GuidedTarget.CenterPosition;
```
This ignores the range limit, which looks rather odd when you target a fast moving vehicle that is running away.  What was the behaviour of the original game in that case?

reaperrr replied:
I think in the original it would stop shooting (not suddenly, fading out) and probably stop tracking and target the last actor position that was in range until the tail reached that position.

I'll check later.

reaperrr commented:
Updated, tracking takes weapon range into consideration now.

pchote commented on OpenRA.Mods.Common/Projectiles/AreaBeam.cs:
```diff
@@ -54,6 +54,9 @@ public class AreaBeamInfo : IProjectileInfo
 		public readonly bool Blockable = false;
 
+		[Desc("Does the beam follow the target.")]
+		public readonly bool TracksTarget = false;
```
`TrackTarget` (no `s`) would be more consistent with the other fields: `Render(s)Beam`, `Use(s)PlayerColor`.

pchote commented on OpenRA.Mods.Common/Projectiles/AreaBeam.cs:
```diff
@@ -132,8 +136,38 @@ public AreaBeam(AreaBeamInfo info, ProjectileArgs args, Color color)
+
+				// Only continue tracking target if it's within weapon range +
+				// 1 cell (1024) leniency to avoid edge case stuttering (start firing and immediately stop again).
+				if (targetDistance > args.Weapon.Range.Length + 1024)
```
Would `BeyondTargetRange` be more appropriate than hardcoding 1 cell?

reaperrr replied:
I'm inclined to say 'yes'.

pchote commented on OpenRA.Mods.Common/Projectiles/AreaBeam.cs:
```diff
@@ -132,8 +136,38 @@ public AreaBeam(AreaBeamInfo info, ProjectileArgs args, Color color)
+			// Update the target position with the range we shoot beyond the target by
+			// I.e. we can deliberately overshoot, so aim for that position
+			var dir = new WVec(0, -1024, 0).Rotate(WRot.FromFacing(towardsTargetFacing));
+			target += dir * info.BeyondTargetRange.Length / 1024;
```
If yes, then we'd want to account for that here so that we don't end up with 2 * `BeyondTargetRange` overshoot.

reaperrr replied:
Are you sure? The check above only checks the distance, it doesn't add `BeyondTargetRange` to the target, and we return if the distance exceeds weapon range + `BeyondTargetRange` so we wouldn't even arrive here if the distance was larger.

pchote replied:
I'm not sure, and you appear to be correct.

pchote replied:
if `!args.GuidedTarget.IsValidFor(args.SourceActor)` then this delta will be added to the already-offset position from the last tick.  I don't think this is what we want?

pchote reviewed (changes requested):
Looks fine once we sort out the following:

reaperrr committed:
Rename LaserZap.TracksTarget to TrackTarget

And fix engineVersion of older Duration upgrade rule.

reaperrr commented:
Updated and rebased.

Additionally, for consistency I renamed `LaserZap`'s `TracksTarget` to `TrackTarget` as well and fixed some projectile upgrade rule dates (they were merged post-prep).

pchote commented on OpenRA.Mods.Common/Projectiles/AreaBeam.cs:
```diff
@@ -132,8 +136,38 @@ public AreaBeam(AreaBeamInfo info, ProjectileArgs args, Color color)
+			if (args.GuidedTarget.IsValidFor(args.SourceActor))
+			{
+				var guidedTargetPos = args.GuidedTarget.CenterPosition;
+				var targetDistance = (guidedTargetPos - args.Source).Length;
```
You're comparing this against other `WDist`s, so you can/should avoid the `.Length` in these.

reaperrr commented:
Updated.

pchote commented on OpenRA.Mods.Common/Projectiles/AreaBeam.cs:
```diff
@@ -132,8 +136,38 @@ public AreaBeam(AreaBeamInfo info, ProjectileArgs args, Color color)
+				if (targetDistance > args.Weapon.Range + info.BeyondTargetRange)
+				{
+					continueTracking = false;
+					isTailTravelling = true;
```
Can you please pull out a `StopTargeting` method that is used here and when the target dies?  This will cut down on the number of places that directly manipulate these flags.

pchote commented on OpenRA.Mods.Common/Projectiles/AreaBeam.cs:
```diff
@@ -132,8 +136,38 @@ public AreaBeam(AreaBeamInfo info, ProjectileArgs args, Color color)
+
 		public void Tick(World world)
 		{
+			if (info.TrackTarget && continueTracking)
```
You can make this a bit cleaner by moving the `continueTracking` check inside `UpdateTarget` (it can tie in with the fix for the issue just above) and then rename the method to `TrackTarget`.  This then keeps all the logic related to target tracking in one place.

reaperrr commented:
Updated.

pchote commented on OpenRA.Mods.Common/Projectiles/AreaBeam.cs:
```diff
@@ -149,13 +189,13 @@ public void Tick(World world)
 			}
 
 			// Allow for 1 cell (1024) leniency to avoid edge case stuttering (start firing and immediately stop again).
```
This is no longer hardcoded to 1 cell.

pchote commented on OpenRA.Mods.Common/UtilityCommands/UpgradeRules.cs:
```diff
@@ -491,19 +491,26 @@ internal static void UpgradeWeaponRules(ModData modData, int engineVersion, ref
 				}
 
+				// Rename LaserZap TracksTarget to TrackTarget
+				if (engineVersion < 20161102)
```
Please update this date.

pchote reviewed (dismissed):
Two minor code nits:

pchote reviewed (commented):
The beam head disconnecting from the target when the tail disconnects from the source looks weird, but I just confirmed that this is consistent with the original game.

pchote commented:
:+1: after fixing those minor nits and rebasing.

reaperrr commented:
Updated and rebased.

obrakmann reviewed (approved):
👍

obrakmann commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/93b5e94506f44f002e3e6012e775ae0fac5f1d52)
